File: Algorithms/Tree-Algorithm/Sorted_Array_To_BST/SortedArrayToBST.java
class Node {
     
    int data;
    Node left, right;
     
    Node(int d) {
        data = d;
        left = right = null;
    }
}
 
class BinaryTree {
 
    Node sortedArrayToBST(int arr[], int start, int end) {
 
        if (start > end) {
            return null;
        }
 
        /* Get the middle element and make it root */
        int mid = (start + end) / 2;
        Node node = new Node(arr[mid]);
 
        //Construct the left subtree and make it
        //left child of root using recursion
        node.left = sortedArrayToBST(arr, start, mid - 1);
 
        //Construct the right subtree and make it
        //right child of root using recursion
        node.right = sortedArrayToBST(arr, mid + 1, end);
         
        return node;
    }
 
    //Print preorder traversal of BST 
    void preOrder(Node node) {
        if (node == null) {
            return;
        }
        System.out.print(node.data + " ");
        preOrder(node.left);
        preOrder(node.right);
    }
}

public class SortedArrayToBST {

	public static void main(String[] args) {
        BinaryTree tree = new BinaryTree();
        int arr[] = new int[]{10, 20, 30, 40, 50, 60, 70};
        int n = arr.length;
        Node root = tree.sortedArrayToBST(arr, 0, n - 1);
        
        System.out.println("Preorder traversal of generated BST");
        tree.preOrder(root);
    }
}

/* 
 * Output:
 * 			Preorder traversal of generated BST
 *			40 20 10 30 60 50 70 
 * 
 */

File: Algorithms/Tree-Algorithm/Sorted-Array-To-BST/Sorted-Array-To-BST.cpp
#include<bits/stdc++.h>
using namespace std;

/* A Binary Tree node */
class TNode
{
	public:
	int data;
	TNode* left;
	TNode* right;
};

TNode* newNode(int data);

/* A function that constructs Balanced
Binary Search Tree from a sorted array */
TNode* sortArrayToBST(int arr[], int start, int end)
{
	/* Base Case */
	if (start > end)
	return NULL;

	/* Get the middle element and make it root */
	int mid = (start + end)/2;
	TNode *root = newNode(arr[mid]);

	/* Recursively construct the left subtree
	and make it left child of root */
	root->left = sortArrayToBST(arr, start, mid - 1);

	/* Recursively construct the right subtree
	and make it right child of root */
	root->right = sortArrayToBST(arr, mid + 1, end);

	return root;
}

/* Helper function that allocates a new node
with the given data and NULL left and right
pointers. */
TNode* newNode(int data)
{
	TNode* node = new TNode();
	node->data = data;
	node->left = NULL;
	node->right = NULL;

	return node;
}

/* A utility function to print
preorder traversal of BST */
void preOrder(TNode* node)
{
	if (node == NULL)
		return;
	cout << node->data << " ";
	preOrder(node->left);
	preOrder(node->right);
}

// Driver Code
int main()
{
	int n;
	cin>>n;
	int arr[n];
	for(i=0;i<n:i++)
	{
	  cin>>arr;
	}

	/* Convert List to BST */
	TNode *root = sortArrayToBST(arr, 0, n-1);
	cout << "PreOrder Traversal of constructed BST \n";
	preOrder(root);

	return 0;
}
